micro_sam/training/joint_sam_trainer.py

```diff
@@ -0,0 +1,214 @@
+        for block, sampler in zip(self.blocks, self.samplers):
+            x = sampler(x)
+            x = block(x)
+
+        return x
+
+
+class UNETRForJointTraining(UNETR):
+    def __init__(
+        self,
+        encoder: Optional[nn.Module] = None,
+        out_channels: int = 1,
+        final_activation: Optional[Union[str, nn.Module]] = None,
+        **kwargs
+    ) -> None:
+        super().__init__(encoder, out_channels, **kwargs)
+
+        self.encoder = encoder
+
+        # parameters for the decoder network
+        depth = 3
+        initial_features = 64
+        gain = 2
+        features_decoder = [initial_features * gain ** i for i in range(depth + 1)][::-1]
+        scale_factors = depth * [2]
+        self.out_channels = out_channels
+
+        self.decoder = Decoder(
+            features=features_decoder,
+            scale_factors=scale_factors[::-1],
+            conv_block_impl=ConvBlock2d,
+            sampler_impl=Upsampler2d
+        )
+
+        self.out_conv = nn.Conv2d(features_decoder[-1], out_channels, 1)
+        self.final_activation = self._get_activation(final_activation)
+
+    def forward(self, x):
+        org_shape = x.shape[-2:]
+
+        x = torch.stack([self.preprocess(e) for e in x], dim=0)
+
+        x = self.encoder(x)
+        x = self.decoder(x)
+
+        x = self.out_conv(x)
+        if self.final_activation is not None:
+            x = self.final_activation(x)
+
+        x = self.postprocess_masks(x, org_shape, org_shape)
+
+        return x
+
+
+class JointSamTrainer(SamTrainer):
+    def __init__(
+            self, **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.unetr = UNETRForJointTraining(
+            img_size=self.model.img_size,
+            backbone="sam",
+            encoder="vit_b",
+            out_channels=self.model.out_channels,
+            use_sam_stats=True,
+            final_activation="Sigmoid"
+        )
+
+    def _instance_train_iteration(self, x, y):
+        # we pass the inputs to the unetr model
+        # get the outputs, calculate the dice loss
+        # return the dice loss
+        instance_loss = ...
+        return instance_loss
+
+    def _train_epoch_impl(self, progress, forward_context, backprop):
+        self.model.train()
+
+        input_check_done = False
+
+        n_iter = 0
+        t_per_iter = time.time()
+        for x, y in self.train_loader:
+            input_check_done = self._check_input_normalization(x, input_check_done)
+
+            self.optimizer.zero_grad()
+
+            with forward_context():
+                # 1. train for the interactive segmentation
+                (loss, mask_loss, iou_regression_loss, model_iou,
+                 sampled_binary_y) = self._interactive_train_iteration(x, y, self._iteration)
+
+            backprop(loss)
+
+            # let's get the unetr decoder for doing the instance segmentation
+            self.unetr.encoder = self.model.encoder  # TODO: revisit
```

Comment:
This should not be necessary if you initialize the unetr correctly.

Reply:
I still need to have a closer look at this. Will keep you posted.